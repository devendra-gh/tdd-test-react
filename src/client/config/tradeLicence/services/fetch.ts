import baseFetch from 'universal-fetch';
import cookie from 'react-cookies';
import baseUrl from 'client/utils/baseUrl';

const defaults = {
  method: 'POST', // handy with GraphQL backends
  mode: 'same-origin',
  credentials: 'same-origin',
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
};

/**
 * @param {string} url
 * @param {string} method
 * @param {Object} data
 * @param {boolean} fileUpload
 * @returns {Promise<any | never>}
 */
export default async function api(
  url: string,
  method: string = 'GET',
  data?: object,
  fileUpload: boolean = false,
) {
  const config = {
    ...defaults,
    method,
    body: {},
  };

  if (method === 'POST' || method === 'PUT') {
    config.headers = {
      ...config.headers,
      // @ts-ignore
      'csrf-token': cookie.load('XSRF-TOKEN'),
    };
  }

  if (data && method !== 'GET') {
    if (!fileUpload) {
      config.headers = {
        ...config.headers,
        Accept: 'application/json',
        'Content-Type': 'application/json',
      };
    }
    if (fileUpload) {
      delete config.headers['Content-Type'];
    }

    config.body = fileUpload ? data : JSON.stringify(data);
  } else {
    delete config.body;
  }

  try {
    const response = await baseFetch(`${baseUrl}${url}`, config);
    const payload = response.json();
    return payload;
  } catch (e) {
    throw e;
  }
}
