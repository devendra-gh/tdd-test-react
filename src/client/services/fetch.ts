import { IVariables } from '@tamm/app-composer';
import baseFetch from 'universal-fetch';
import cookie from 'react-cookies';
import baseUrl from '../utils/baseUrl';

interface IFetchConfig {
  method: string;
  mode: string;
  credentials: string;
  headers: IVariables;
}

const defaults: IFetchConfig = {
  method: 'POST', // handy with GraphQL backends
  mode: 'same-origin',
  credentials: 'same-origin',
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    'X-Requested-With': 'XMLHttpRequest',
  },
};
/**
 * @param {string} url
 * @param {string} method
 * @param {Object} data
 * @param {boolean} fileUpload
 * @returns {Promise<any | never>}
 */
export default async function api(
  url: string,
  method: string = 'GET',
  data?: object,
  fileUpload: boolean = false,
) {
  // eslint-disable-next-line no-param-reassign
  method = method.toUpperCase();
  const config = {
    ...defaults,
    method,
    body: {},
  };
  if (method === 'POST' || method === 'PUT') {
    config.headers = {
      ...config.headers,
      'csrf-token': cookie.load('XSRF-TOKEN'),
    };
  }
  if (data && method !== 'GET') {
    if (!fileUpload) {
      config.headers = {
        ...config.headers,
        Accept: 'application/json',
        'Content-Type': 'application/json',
      };
    }
    if (fileUpload) {
      delete config.headers['Content-Type'];
    }
    config.body = fileUpload ? data : JSON.stringify(data);
  } else {
    delete config.body;
  }
  try {
    let response;
    if (url.includes('https://')) response = await baseFetch(url, data);
    else response = await baseFetch(`${baseUrl}${url}`, config);
    const payload = response.json();
    return payload;
  } catch (e) {
    throw e;
  }
}
