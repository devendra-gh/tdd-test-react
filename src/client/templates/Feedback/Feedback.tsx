import React, { useCallback, useState } from 'react';

import GlobalCsatFeedback from '@tamm/ui-lib-v2-global-csat-feedback';

import fetch from 'client/services/fetch';

interface FeedbackOptions {
  ease: number;
  quality: number;
  performance: number;
  comment: string;
}

interface FeedbackProps {
  i18n: any;
}

export function Feedback(props: FeedbackProps) {
  const [status, setStatus] = useState<
    'idle' | 'active' | 'pending' | 'success' | 'error'
  >('idle');
  const handleSubmit = useCallback(
    async (smileyType: string, options: FeedbackOptions) => {
      setStatus('pending');
      try {
        await fetch('/pub/adfeedback/feedbacks', 'POST', {
          surveyType: 'general',
          smileyType,
          comments: options.comment,
          ratings: {
            ease: options.ease,
            quality: options.quality,
            performance: options.performance,
          },
          pageUrl: window.location.href,
        });
        setStatus('success');
      } catch (e) {
        setStatus('error');
      }
    },
    [],
  );

  return (
    <GlobalCsatFeedback
      onSubmit={handleSubmit}
      i18n={props.i18n}
      status={status}
      autoOpen={false}
    />
  );
}
