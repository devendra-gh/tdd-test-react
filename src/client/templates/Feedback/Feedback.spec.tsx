import React from 'react';
import { shallow } from 'enzyme';
import GlobalCsatFeedback from '@tamm/ui-lib-v2-global-csat-feedback';
import fetch from 'client/services/fetch';
import { Feedback } from './Feedback';

jest.mock('client/services/fetch');

describe('Feedback', () => {
  it('should render component', () => {
    const component = shallow(<Feedback i18n={() => ({})} />);
    expect(component).toMatchSnapshot();
  });

  it('should render success state', async () => {
    const component = shallow(<Feedback i18n={() => ({})} />);
    (fetch as jest.Mock).mockResolvedValue({});
    const onSubmit =
      component.find(GlobalCsatFeedback).prop('onSubmit') || (() => {});
    await onSubmit('test', {});
    expect(component).toMatchSnapshot();
  });

  it('should render error state', async () => {
    const component = shallow(<Feedback i18n={() => ({})} />);
    (fetch as jest.Mock).mockRejectedValue({});
    const onSubmit =
      component.find(GlobalCsatFeedback).prop('onSubmit') || (() => {});
    await onSubmit('test', {});
    expect(component).toMatchSnapshot();
  });
});
