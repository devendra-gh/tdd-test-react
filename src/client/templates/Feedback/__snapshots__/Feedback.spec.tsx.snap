// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback should render component 1`] = `
<GlobalCsatFeedback
  autoOpen={false}
  containerId=""
  i18n={[Function]}
  isOpen={false}
  onSubmit={[Function]}
  shouldCloseOnEsc={false}
  shouldCloseOnOverlayClick={false}
  status="idle"
  withComment={true}
/>
`;

exports[`Feedback should render error state 1`] = `
<GlobalCsatFeedback
  autoOpen={false}
  containerId=""
  i18n={[Function]}
  isOpen={false}
  onSubmit={[Function]}
  shouldCloseOnEsc={false}
  shouldCloseOnOverlayClick={false}
  status="error"
  withComment={true}
/>
`;

exports[`Feedback should render success state 1`] = `
<GlobalCsatFeedback
  autoOpen={false}
  containerId=""
  i18n={[Function]}
  isOpen={false}
  onSubmit={[Function]}
  shouldCloseOnEsc={false}
  shouldCloseOnOverlayClick={false}
  status="success"
  withComment={true}
/>
`;
